site-content/source/modules/ROOT/pages/blog/Apache-Cassandra-4.1-Features-Guardrails-Framework.adoc

```diff
@@ -0,0 +1,134 @@
+```
+
+The triggering of a guardrail will always emit a diagnostic event of type `GuardrailEvent`. Thus, anyone subscribing to diagnostic events of that type will be able to monitor guardrail violations and trigger any desired actions. Some examples of actions include storing metrics and contacting the offending user, etc.
+
+=== Users and Extensibility
+
+Guardrails are only applied to the operations of regular users, so they will neither be checked for superuser queries nor internal queries. By default, all regular users are subjected to the same guardrails configuration values defined in `cassandra.yaml`. However, the configuration for guardrails is an extensible API defined by the interfaces `GuardrailsConfig` and `GuardrailsConfigProvider`. These interfaces provide the configurations of every guardrail as a function of the user running the guarded operation. Third-party alternative implementations could provide different guardrail configurations depending on the user, or on some other factors. Custom implementations can be provided through the JVM system property `cassandra.custom_guardrails_config_provider_class`. It is important to know that this API isn’t officially supported yet, and there can be changes breaking backward compatibility in any minor release.
+
+=== Background Guardrails
+
+In general, guardrails are associated with a specific CQL query. However, due to technical limitations, some guardrails are checked in the background, without being associated with any specific query. That’s the case, for example, with the guardrails for the size or number of items of a https://cassandra.apache.org/doc/trunk/cassandra/cql/types.html#collections[non-frozen collection]. Although we can do some checks when a query writes a new collection fragment, we cannot know if there are other fragments of the collection previously stored on the https://cassandra.apache.org/doc/trunk/cassandra/architecture/storage_engine.html#sstables[SSTables]. We could, of course, check into the SSTables, but that would involve a costly read-before-write operation. Instead, the guardrail checks the size of all collections every time an SSTable is written to disk, such as on https://cassandra.apache.org/doc/trunk/cassandra/architecture/storage_engine.html#memtables[memtable] flush or during https://cassandra.apache.org/doc/trunk/cassandra/operating/compaction/index.html[compaction]. If a large collection is detected the guardrail is triggered and emits the proper log messages and diagnostic events. However, in contrast to what happens with regular guardrails, we won’t abort any operation. The reason for this is that SSTable writes aren't associated with any specific CQL query. Rather, they are the result of multiple past CQL writes that have already happened long before. Future guardrails for similar things, like partition size, are likely to work in the same way.
```

Comment:
Replace hardcoded `/doc` URLs with `link:`:
```suggestion
In general, guardrails are associated with a specific CQL query. However, due to technical limitations, some guardrails are checked in the background, without being associated with any specific query. That’s the case, for example, with the guardrails for the size or number of items of a link:/doc/trunk/cassandra/cql/types.html#collections[non-frozen collection^]. Although we can do some checks when a query writes a new collection fragment, we cannot know if there are other fragments of the collection previously stored on the link:/doc/trunk/cassandra/architecture/storage_engine.adoc#sstables[SSTables^]. We could, of course, check into the SSTables, but that would involve a costly read-before-write operation. Instead, the guardrail checks the size of all collections every time an SSTable is written to disk, such as on link:/doc/trunk/cassandra/architecture/storage_engine.html#memtables[memtable^] flush or during link:/doc/trunk/cassandra/operating/compaction/index.adoc[compaction^]. If a large collection is detected the guardrail is triggered and emits the proper log messages and diagnostic events. However, in contrast to what happens with regular guardrails, we won’t abort any operation. The reason for this is that SSTable writes aren't associated with any specific CQL query. Rather, they are the result of multiple past CQL writes that have already happened long before. Future guardrails for similar things, like partition size, are likely to work in the same way.
```

Reply:
Oh, I missed the `link:` prefix on my suggestions about relative paths above.